components/score_tracker.py

```diff
@@ -0,0 +1,150 @@
+        self.state_red = np.zeros((3, 9), dtype=bool)
+        self.state_blue = np.zeros((3, 9), dtype=bool)
+        self.did_states_change = magicbot.will_reset_to(False)
+
+    def add_vision_data(
+        self, side: wpilib.DriverStation.Alliance, pos: npt.ArrayLike, confidence: float
+    ) -> None:
+        confidences = (
+            self.confidences_red if side == side.kRed else self.confidences_blue
+        )
+        confidences[pos] = confidences[
+            pos
+        ] * ScoreTracker.CONF_EXP_FILTER_ALPHA + confidence * (
+            1.0 - ScoreTracker.CONF_EXP_FILTER_ALPHA
+        )
+        if abs(confidences[pos]) > ScoreTracker.CONF_THRESHOLD:
+            self.did_states_change = True
+            state = self.state_red if side == side.kRed else self.state_blue
+            state[pos] = confidence > 0.0
+
+    @staticmethod
+    def count_links(r: npt.NDArray[bool]) -> int:
+        i = 0
+        n = 0
+        l = len(r)
+        while i < l - 2:
+            if r[i] and r[i + 1] and r[i + 2]:
+                n += 1
+                i += 3
+                continue
+            i += 1
+        return n
+
+    @staticmethod
+    def evaluate_state(a: npt.NDArray[bool]) -> int:
+        return (
+            sum(ScoreTracker.count_links(r) for r in a) * 5
+            + a[0].sum() * 5
+            + a[1].sum() * 3
+            + a[2].sum() * 2
+        )
+
+    @staticmethod
+    def run_lengths_mod3(state: npt.NDArray[bool]) -> npt.NDArray[int]:
+        """
+        Returns an array where corresponding in shape to the input, where
+        every value is replaced by the length of the longest uninterrupted
+        run of true values containing it, modulo 3
+        """
+        run_lengths = np.zeros_like(state, dtype=int)
+        for y in range(3):
+            x = 0
+            while x < 9:
+                if not state[y, x]:
+                    x += 1
+                    continue
+                acc = 0
+                for xn in range(x, 9):
+                    if not state[y, xn]:
+                        break
+                    acc += 1
+                run_lengths[y, x : x + acc] = acc % 3
+                x += acc
+        return run_lengths
+
+    @staticmethod
+    def get_in_row(arr: npt.NDArray, x: int, y: int, def_val):
+        if x < 0 or x > 8:
+            return def_val
+        else:
+            return arr[y, x]
+
+    @staticmethod
+    def get_best_moves(
+        state: npt.NDArray[bool],
+        type_to_test: GridNode,
+        link_preparation_score: float = 2.5,
+    ) -> npt.NDArray:
+        vals = np.zeros_like(state, dtype=float)
+        run_lengths = ScoreTracker.run_lengths_mod3(state)
+        for y in range(3):
+            for x in range(9):
+                if (
+                    state[y, x]
+                    or (
+                        type_to_test == GridNode.CUBE
+                        and not ScoreTracker.CUBE_MASK[y, x]
```

Comment:
This would be better off as a class method rather than a static method, referring to `cls` instead of the class name.